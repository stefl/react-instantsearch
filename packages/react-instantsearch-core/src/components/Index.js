import React, { Component, Children } from 'react';
import PropTypes from 'prop-types';

/* eslint valid-jsdoc: 0 */
/**
 * @description
 * `<Index>` is the component that allows you to apply widgets to a dedicated index. It's
 * useful if you want to build an interface that targets multiple indices.
 * @kind widget
 * @name <Index>
 * @propType {string} indexName - index in which to search.
 * @propType {{ Root: string|function, props: object }} [root] - Use this to customize the root element. Default value: `{ Root: 'div' }`
 * @example
 * import React from 'react';
 * import { InstantSearch, Index, SearchBox, Hits, Configure } from 'react-instantsearch-dom';
 *
 * const App = () => (
 *   <InstantSearch
 *     appId="latency"
 *     apiKey="6be0576ff61c053d5f9a3225e2a90f76"
 *     indexName="ikea"
 *   >
 *     <Configure hitsPerPage={5} />
 *     <SearchBox />
 *     <Index indexName="ikea">
 *       <Hits />
 *     </Index>
 *     <Index indexName="bestbuy">
 *       <Hits />
 *     </Index>
 *   </InstantSearch>
 * );
 */
class Index extends Component {
  constructor(props, context) {
    super(props);
    const {
      ais: { widgetsManager },
    } = context;

    /*
     we want <Index> to be seen as a regular widget.
     It means that with only <Index> present a new query will be sent to Algolia.
     That way you don't need a virtual hits widget to use the connectAutoComplete.
    */
    this.unregisterWidget = widgetsManager.registerWidget(this);
  }

  componentWillMount() {
    this.context.ais.onSearchParameters(
      this.getSearchParameters,
      this.getChildContext(),
      this.props
    );
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.indexName !== nextProps.indexName) {
      this.context.ais.widgetsManager.update();
    }
  }

  componentWillUnmount() {
    this.unregisterWidget();
  }

  getChildContext() {
    return {
      multiIndexContext: {
        targetedIndex: this.props.indexName,
      },
    };
  }

  getSearchParameters(searchParameters, props) {
    return searchParameters.setIndex(
      this.props ? this.props.indexName : props.indexName
    );
  }

  render() {
    const childrenCount = Children.count(this.props.children);
    const { Root, props } = this.props.root;
    if (childrenCount === 0) return null;
    else return <Root {...props}>{this.props.children}</Root>;
  }
}

Index.propTypes = {
  // @TODO: These props are currently constant.
  indexName: PropTypes.string.isRequired,
  children: PropTypes.node,
  root: PropTypes.shape({
    Root: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
    props: PropTypes.object,
  }).isRequired,
};

Index.childContextTypes = {
  multiIndexContext: PropTypes.object.isRequired,
};

Index.contextTypes = {
  // @TODO: more precise widgets manager propType
  ais: PropTypes.object.isRequired,
};

export default Index;
