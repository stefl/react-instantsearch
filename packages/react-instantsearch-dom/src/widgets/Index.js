import { createIndex } from 'react-instantsearch-core';

const Index = createIndex({
  Root: 'div',
  props: {
    className: 'ais-MultiIndex__root',
  },
});

export default Index;
