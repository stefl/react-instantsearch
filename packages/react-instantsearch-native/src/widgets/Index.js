import { View } from 'react-native';
import { createIndex } from 'react-instantsearch-core';

const Index = createIndex({
  Root: View,
});

export default Index;
