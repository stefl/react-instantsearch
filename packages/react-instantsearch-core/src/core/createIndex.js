import React from 'react';
import PropTypes from 'prop-types';
import Index from '../components/Index';

/**
 * Creates a specialized root Index component. It accepts
 * a specification of the root Element.
 * @param {object} defaultRoot - the defininition of the root of an Index sub tree.
 * @return {object} a Index root
 */
const createIndex = defaultRoot => {
  const CreateIndex = ({ indexName, root, children }) => (
    <Index indexName={indexName} root={root}>
      {children}
    </Index>
  );

  CreateIndex.propTypes = {
    indexName: PropTypes.string.isRequired,
    root: PropTypes.shape({
      Root: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
      props: PropTypes.object,
    }),
    children: PropTypes.node,
  };

  CreateIndex.defaultProps = {
    root: defaultRoot,
  };

  return CreateIndex;
};

export default createIndex;
