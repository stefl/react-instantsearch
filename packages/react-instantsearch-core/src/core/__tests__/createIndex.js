import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Index from '../../components/Index';
import createIndex from '../createIndex';

Enzyme.configure({ adapter: new Adapter() });

describe('createIndex', () => {
  const CustomIndex = createIndex({ Root: 'div' });

  it('wraps Index', () => {
    const wrapper = shallow(<CustomIndex indexName="name" />);
    expect(wrapper.is(Index)).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('expect to create Index with a custom root props', () => {
    const root = {
      Root: 'span',
      props: {
        style: {
          flex: 1,
        },
      },
    };

    const wrapper = shallow(<CustomIndex indexName="name" root={root} />);

    expect(wrapper.is(Index)).toBe(true);
    expect(wrapper.props().root).toEqual(root);
    expect(wrapper).toMatchSnapshot();
  });
});
