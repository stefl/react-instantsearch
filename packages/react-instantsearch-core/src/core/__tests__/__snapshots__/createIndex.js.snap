// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createIndex expect to create Index with a custom root props 1`] = `
<Index
  indexName="name"
  root={
    Object {
      "Root": "span",
      "props": Object {
        "style": Object {
          "flex": 1,
        },
      },
    }
  }
/>
`;

exports[`createIndex wraps Index 1`] = `
<Index
  indexName="name"
  root={
    Object {
      "Root": "div",
    }
  }
/>
`;
